From fd2cfb15bf362c66db03ba4ceed93c6199b07c9f Mon Sep 17 00:00:00 2001
From: Hou Qi <qi.hou@nxp.com>
Date: Wed, 11 Oct 2023 12:44:47 +0800
Subject: [PATCH 50/75] LF-10264 v4l2: drop frames for decoded buffer with
 error flag

For streams with some frames that cannot be decoded, these frames'
timestamps should not participate in sorting and should be dropped.
Otherwise, it may cause pts of capture buffer smaller than the real
pts, and further cause AV unsync.

Also, imx95 return all capture buffers and some of them may have
error flag. Need to drop these buffers to avoid being showed and
pass below test.

Enter video-test-suite/subtree/fluster, run
./fluster.py run -ts JCT-VC-HEVC_V1 -tv BUMPING_A_ericsson_1 -d GStreamer-H.265-V4L2-Gst1.0 -j1

Upstream-Status: Pending
Part-of: <https://gitlab.freedesktop.org/gstreamer/gstreamer/-/merge_requests/5479>
Signed-off-by: Xiaolei Wang <xiaolei.wang@windriver.com>
---
 sys/v4l2/gstv4l2allocator.c |  8 ++++++++
 sys/v4l2/gstv4l2object.c    |  7 +++++++
 sys/v4l2/gstv4l2object.h    |  2 ++
 sys/v4l2/gstv4l2videodec.c  | 25 +++++++++++++++++++++++++
 4 files changed, 42 insertions(+)

diff --git a/sys/v4l2/gstv4l2allocator.c b/sys/v4l2/gstv4l2allocator.c
index 73da324..c4058f9 100644
--- a/sys/v4l2/gstv4l2allocator.c
+++ b/sys/v4l2/gstv4l2allocator.c
@@ -1435,6 +1435,8 @@ gst_v4l2_allocator_dqbuf (GstV4l2Allocator * allocator,
   if (V4L2_TYPE_IS_OUTPUT (obj->type)) {
     gst_v4l2_allocator_reset_size (allocator, group);
     if (group->buffer.flags & V4L2_BUF_FLAG_ERROR) {    //output buffer cannot be decoded
+      obj->drop_frames =
+          g_list_append (obj->drop_frames, &group->buffer.timestamp);
       if (!obj->frame_decoded)
         obj->err_cnt++;
     } else {
@@ -1442,6 +1444,12 @@ gst_v4l2_allocator_dqbuf (GstV4l2Allocator * allocator,
       obj->frame_decoded = TRUE;
     }
   } else {
+    if (group->buffer.flags & V4L2_BUF_FLAG_ERROR) {    //capture buffer has error flag
+      GST_INFO_OBJECT (allocator, "CAPTURE error flag for frame %u",
+          (guint32) group->buffer.timestamp.tv_sec);
+      obj->drop_frames =
+          g_list_append (obj->drop_frames, &group->buffer.timestamp);
+    }
     /* for capture, simply read the size */
     for (i = 0; i < group->n_mem; i++) {
       gsize size, offset;
diff --git a/sys/v4l2/gstv4l2object.c b/sys/v4l2/gstv4l2object.c
index 9ac7043..ef825de 100644
--- a/sys/v4l2/gstv4l2object.c
+++ b/sys/v4l2/gstv4l2object.c
@@ -566,6 +566,7 @@ gst_v4l2_object_new (GstElement * element,
   v4l2object->norms = NULL;
   v4l2object->channels = NULL;
   v4l2object->colors = NULL;
+  v4l2object->drop_frames = NULL;
 
   v4l2object->drm_modifier = 0;
   v4l2object->is_g2 = FALSE;
@@ -627,6 +628,7 @@ gst_v4l2_object_destroy (GstV4l2Object * v4l2object)
   g_free (v4l2object->videodev);
   g_free (v4l2object->par);
   g_free (v4l2object->channel);
+  g_free (v4l2object->drop_frames);
 
   gst_poll_free (v4l2object->poll);
 
@@ -1026,6 +1028,11 @@ gst_v4l2_object_close (GstV4l2Object * v4l2object)
   if (v4l2object->poll)
     gst_poll_remove_fd (v4l2object->poll, &v4l2object->pollfd);
 
+  if (v4l2object->drop_frames) {
+    g_free (v4l2object->drop_frames);
+    v4l2object->drop_frames = NULL;
+  }
+
   return TRUE;
 }
 
diff --git a/sys/v4l2/gstv4l2object.h b/sys/v4l2/gstv4l2object.h
index eaa79c4..f1ad009 100644
--- a/sys/v4l2/gstv4l2object.h
+++ b/sys/v4l2/gstv4l2object.h
@@ -200,6 +200,8 @@ struct _GstV4l2Object {
   GList *channels;
   GData *controls;
 
+  GList *drop_frames;
+
   /* properties */
   v4l2_std_id tv_norm;
   gchar *channel;
diff --git a/sys/v4l2/gstv4l2videodec.c b/sys/v4l2/gstv4l2videodec.c
index b44cbd9..3cfe668 100644
--- a/sys/v4l2/gstv4l2videodec.c
+++ b/sys/v4l2/gstv4l2videodec.c
@@ -914,6 +914,15 @@ gst_v4l2_video_dec_loop (GstVideoDecoder * decoder)
       ret = gst_v4l2_buffer_pool_process (cpool, &buffer, NULL);
       if (cpool)
         gst_object_unref (cpool);
+
+      while (self->v4l2capture->drop_frames) {
+        guint32 *tmp = self->v4l2capture->drop_frames->data;
+        GstVideoCodecFrame *frame = gst_video_decoder_get_frame (decoder, *tmp);
+        if (frame)
+          gst_video_decoder_drop_frame (decoder, frame);
+        self->v4l2capture->drop_frames =
+            g_list_remove (self->v4l2capture->drop_frames, tmp);
+      }
     }
   } while (ret == GST_V4L2_FLOW_CORRUPTED_BUFFER);
 
@@ -1118,6 +1127,14 @@ gst_v4l2_video_dec_handle_frame (GstVideoDecoder * decoder,
     ret =
         gst_v4l2_buffer_pool_process (GST_V4L2_BUFFER_POOL (pool), &codec_data,
         processed ? &frame->system_frame_number : &dummy_frame_number);
+    while (self->v4l2output->drop_frames) {
+      guint32 *tmp = self->v4l2output->drop_frames->data;
+      GstVideoCodecFrame *frame = gst_video_decoder_get_frame (decoder, *tmp);
+      if (frame)
+        gst_video_decoder_drop_frame (decoder, frame);
+      self->v4l2output->drop_frames =
+          g_list_remove (self->v4l2output->drop_frames, tmp);
+    }
     GST_VIDEO_DECODER_STREAM_LOCK (decoder);
 
     gst_buffer_unref (codec_data);
@@ -1158,6 +1175,14 @@ gst_v4l2_video_dec_handle_frame (GstVideoDecoder * decoder,
     ret =
         gst_v4l2_buffer_pool_process (GST_V4L2_BUFFER_POOL (pool),
         &frame->input_buffer, &frame->system_frame_number);
+    while (self->v4l2output->drop_frames) {
+      guint32 *tmp = self->v4l2output->drop_frames->data;
+      GstVideoCodecFrame *frame = gst_video_decoder_get_frame (decoder, *tmp);
+      if (frame)
+        gst_video_decoder_drop_frame (decoder, frame);
+      self->v4l2output->drop_frames =
+          g_list_remove (self->v4l2output->drop_frames, tmp);
+    }
     GST_VIDEO_DECODER_STREAM_LOCK (decoder);
 
     if (self->v4l2output->err_cnt > MAX_OUTPUT_ERROR_COUNT) {
-- 
2.43.0

